narwhal/primary/src/core.rs

```diff
@@ -608,19 +610,30 @@ impl Core {
             .certificates_processed
             .with_label_values(&[certificate_source])
             .inc();
+
         // Append the certificate to the aggregator of the
         // corresponding round.
-        self.append_certificate_in_aggregator(certificate.clone())
-            .await?;
+        let digest = certificate.digest();
+        if let Err(e) = self
+            .append_certificate_in_aggregator(certificate.clone())
+            .await
+        {
+            warn!(
+                "Failed to aggregate certificate {} for header: {}",
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
 
         // Send it to the consensus layer.
-        let digest = certificate.header.digest();
         if let Err(e) = self.tx_new_certificates.send(certificate).await {
             warn!(
                 "Failed to deliver certificate {} to the consensus: {}",
                 digest, e
             );
+            return Err(DagError::ShuttingDown);
```

Comment:
good point!

Reply:
I hope later on we can verify the system is indeed shutting down.